### PR TITLE
Disallow remove last master in cluster (bsc#1138467)

### PR DESCRIPTION
**Pull request by c3y1huang**
Signed-off-by: Chin-Ya Huang <chin-ya.huang@suse.com>
Fixes #1138467

## What does this PR do?

Added checks so `skuba` does not let user remove the last master of the cluster.

### CODE REVIEW

File: pkg/skuba/actions/node/remove/remove.go [hunk 1]
```diff
@@ -39,7 +40,22 @@ func Remove(target string, drainTimeout time.Duration) error {
 
 	node, err := client.CoreV1().Nodes().Get(target, metav1.GetOptions{})
 	if err != nil {
-		return errors.Wrapf(err, "[remove-node] could not get node %s", target)
+		return errors.Wrapf(err, "could not get node %s", target)
+	}
+
+	_, isMaster := node.ObjectMeta.Labels[kubeadmconstants.LabelNodeRoleMaster]
+
```
Comment (innobead) [inline pkg/skuba/actions/node/remove/remove.go]: Please remove the empty line.

File: pkg/skuba/actions/node/remove/remove.go [hunk 2]
```diff
@@ -39,7 +40,22 @@ func Remove(target string, drainTimeout time.Duration) error {
 
 	node, err := client.CoreV1().Nodes().Get(target, metav1.GetOptions{})
 	if err != nil {
-		return errors.Wrapf(err, "[remove-node] could not get node %s", target)
+		return errors.Wrapf(err, "could not get node %s", target)
```
Comment (innobead) [inline pkg/skuba/actions/node/remove/remove.go]: Keep [remove-node] , this is the current log styling in skuba. Please also fix others.

File: pkg/skuba/actions/node/remove/remove.go [hunk 3]
```diff
@@ -31,17 +33,27 @@ import (
...
-func Remove(target string, drainTimeout time.Duration) error {
-	client, err := kubernetes.GetAdminClientSet()
-	if err != nil {
-		return errors.Wrap(err, "unable to get admin client set")
-	}
-
+func Remove(client clientset.Interface, target string, drainTimeout time.Duration) error {
 	node, err := client.CoreV1().Nodes().Get(target, metav1.GetOptions{})
 	if err != nil {
 		return errors.Wrapf(err, "[remove-node] could not get node %s", target)
```
Comment (jhsiaosuse) [inline pkg/skuba/actions/node/remove/remove.go]: if the target node does not exist, what's the command line output to the user?

File: pkg/skuba/actions/node/remove/remove_test.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	}{
+		{
+			name:			"remove last master from cluster",
+			target: 		"master-1",
+			clientset:		fake.NewSimpleClientset(&v1.NodeList{Items: []v1.Node{node1}}),
+			errorExpected: 	true,
+			errorMessage: 	"could not remove last master of the cluster",
+		},
+	}
+
```
Comment (jhsiaosuse) [inline pkg/skuba/actions/node/remove/remove_test.go]: One more test cases could add
```
{
	name:          "cannot get node",
	target:        "master-1",
	clientset:     fake.NewSimpleClientset(),
	errorExpected: true,
	errorMessage:  "[remove-node] could not get node master-1: nodes \"master-1\" not found",
},
```

File: pkg/skuba/actions/node/remove/remove.go [hunk 5]
```diff
@@ -31,17 +33,27 @@ import (
...
 	}
 
+	_, isMaster := node.ObjectMeta.Labels[kubeadmconstants.LabelNodeRoleMaster]
+	if isMaster {
```
Comment (ereslibre) [inline pkg/skuba/actions/node/remove/remove.go]: Minor: you can use `kubernetes.isControlPlane()` helper

File: pkg/skuba/actions/node/remove/remove.go [hunk 6]
```diff
@@ -31,17 +33,27 @@ import (
...
+		nodes, err := client.CoreV1().Nodes().List(metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=", kubeadmconstants.LabelNodeRoleMaster),
+		})
+
+		if err != nil {
+			return errors.Wrapf(err, "could not retrieve master node list")
+		}
+
+		if len(nodes.Items) < 2 {
```
[the remark this answers is not in the document]
Comment (ereslibre) [reply pkg/skuba/actions/node/remove/remove.go]: I think `== 1` is a bit more clear for this case :)

File: pkg/skuba/actions/node/remove/remove.go (hunk 3 above)
Comment (c3y1huang) [reply pkg/skuba/actions/node/remove/remove.go]:
```bash
** This is a BETA release and NOT intended for production usage. **
F0722 09:47:23.559485   16812 remove.go:52] error removing node master-100: [remove-node] could not get node master-100: nodes "master-100" not found
```

**Comment by jhsiaosuse** [comment]
BTW, the description contains `Fixes #`. Need to link a bugzilla or a Github issue?

**Comment by innobead** [comment]
> BTW, the description contains `Fixes #`. Need to link a bugzilla or a Github issue?

Yes, please add bug number in PR name, and also mention bug in description.

**Comment by jhsiaosuse** [comment]
CI `jenkins/skuba-code-lint` failed, please check the import ordering (VScode default `goreturns` will help you correct it or `goimports` does it also)

**Pull request merged**